### ISSUE TITLE
load test: Analytics integration client error on postscore 

### ISSUE THREAD

**Issue by navalev** [MicrosoftDX/nether#397]
#111
while running leaderboard load testing on a Free tier web app, post score reported the following error:
reported when running 100 users loads
```
2017-03-12 14:40:33.969 +00:00 [Error] Error sending analytics ScoreEvent: "System.Net.Sockets.SocketException: An attempt was made to access a socket in a way forbidden by its access permissions
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.EventHubs.Amqp.AmqpEventHubClient.<CreateConnectionAsync>d__31.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.Azure.Amqp.FaultTolerantAmqpObject`1.<OnCreateAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.Azure.Amqp.Singleton`1.<CreateValue>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.Azure.Amqp.Singleton`1.<GetOrCreateAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.EventHubs.Amqp.AmqpEventDataSender.<CreateLinkAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.Azure.Amqp.FaultTolerantAmqpObject`1.<OnCreateAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.Azure.Amqp.Singleton`1.<CreateValue>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.Azure.Amqp.Singleton`1.<GetOrCreateAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.EventHubs.Amqp.AmqpEventDataSender.<OnSendAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.EventHubs.Amqp.AmqpEventDataSender.<OnSendAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.EventHubs.EventHubClient.<SendAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Nether.Integration.Default.Analytics.AnalyticsIntegrationEventHubClient.<SendGameEventAsync>d__2.MoveNext() in C:\Projects\nether\src\Nether.Integration.Default\Analytics\AnalyticsIntegrationEventHubClient.cs:line 30
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Nether.Web.Features.Leaderboard.ScoreController.<SendScoreEventAndLogErrors>d__5.MoveNext() in C:\Projects\nether\src\Nether.Web\Features\Leaderboard\Controllers\ScoreController.cs:line 88"
2017-03-12 14:40:33.969 +00:00 [Information] Executing HttpStatusCodeResult, setting HTTP status code 200
2017-03-12 14:40:33.969 +00:00 [Information] Executed action "Nether.Web.Features.Leaderboard.ScoreController.Post (Nether.Web)" in 1530.0597ms
2017-03-12 14:40:33.969 +00:00 [Information] Request finished in 1532.8306ms 200 
2017-03-12 14:40:33.983 +00:00 [Information] Request starting HTTP/1.1 GET http://nether312.azurewebsites.net/identity/.well-known/openid-configuration/jwks  
2017-03-12 14:40:33.983 +00:00 [Information] Invoking IdentityServer endpoint: "IdentityServer4.Endpoints.DiscoveryEndpoint" for "/.well-known/openid-configuration/jwks"
2017-03-12 14:40:34.172 +00:00 [Information] Successfully validated the token.
2017-03-12 14:40:34.172 +00:00 [Information] HttpContext.User merged via AutomaticAuthentication from authenticationScheme: "Bearer".
2017-03-12 14:40:34.172 +00:00 [Information] AuthenticationScheme: "Bearer" was successfully authenticated.
2017-03-12 14:40:34.172 +00:00 [Information] Scopes found on current principal: "scope: nether-all"
2017-03-12 14:40:34.172 +00:00 [Information] Authorization was successful for user: "loadUser9157".
2017-03-12 14:40:34.172 +00:00 [Information] Executing action method "Nether.Web.Features.Leaderboard.ScoreController.Post (Nether.Web)" with arguments (["Nether.Web.Features.Leaderboard.Models.Score.ScorePostRequestModel"]) - ModelState is Valid
2017-03-12 14:40:34.280 +00:00 [Information] Successfully validated the token.
2017-03-12 14:40:34.280 +00:00 [Information] HttpContext.User merged via AutomaticAuthentication from authenticationScheme: "Bearer".
2017-03-12 14:40:34.280 +00:00 [Information] AuthenticationScheme: "Bearer" was successfully authenticated.
2017-03-12 14:40:34.280 +00:00 [Information] Scopes found on current principal: "scope: nether-all"
2017-03-12 14:40:34.280 +00:00 [Information] Authorization was successful for user: "loadUser9551".
2017-03-12 14:40:34.280 +00:00 [Information] Executing action method "Nether.Web.Features.Leaderboard.ScoreController.Post (Nether.Web)" with arguments (["Nether.Web.Features.Leaderboard.Models.Score.ScorePostRequestModel"]) - ModelState is Valid
```

**Comment by stuartleeks** [comment]
Does this happen for every score that is posted?
Would be interesting to know if this happens on Basic tier.

**Comment by navalev** [comment]
just tested with 25 users on a free tier - happens as well.

**Comment by stuartleeks** [comment]
I've just created https://github.com/MicrosoftDX/nether/pull/402 which changes the config so that the AnalyticsIntegrationEventHubClient instance is reused, rather than created each time it is needed. Since that class maintains the EventHubClient instance as a member, that will cut down on the number of instances of EventHubClient being created
I have a recollection that the EventHubClient internals re-use message factories, but am not 100% sure so it would be interesting to test.

**Comment by navalev** [comment]
testing with #402:
100 users on basic tier - no errors
100 users on free tier - no errors

**Comment by navalev** [comment]
resolved with #402
